File: project.pl
%GENERAL
%convertBinToDec/2
%Needed in Bin to Dec calculation

exp(_,0,1).
exp(X,Y,Z):-
	Y > 0,
	Y1 is Y - 1,
	exp(X,Y1,Z1),
	Z is Z1*X.

convertBinToDec(Bin,Dec):-
	number_string(Bin,S), %convert number to String
	string_chars(S,L), %convert String into list of characters
	length(L,N), %acquire length of string to check index of each number
	N1 is N-1, %subtract length by 1 to get index of the MSB
	convertBinToDec(L,Dec,N1). %recursion on List (LETS GO)
convertBinToDec([H|T],Dec,N):-
	atom_number(H,X), % convert the atom (char) into int
	exp(2,N,B),
	N1 is N -1,
	convertBinToDec(T,Dec1,N1),
	Dec is Dec1 + B*X.

convertBinToDec([],0,_).

%replaceIthItem/4

replaceIthItem(X,[H|T],I,R):-
	replaceIthItem(X,[H|T],0,I,R).
replaceIthItem(X,[_|T],I,I,[X|T]).
replaceIthItem(X,[H1|T1],Count,I,[H2|T2]):-
	H1 \= X,
	H1 = H2,
	Count1 is Count + 1,
	replaceIthItem(X,T1,Count1,I,T2).

%splitEvery/3

splitEvery(_,[],[]).
splitEvery(N, [H|T], [X|T1]):-
	append(X,Y,[H|T]), %append can be used to get more than 1 combination
	length(X,N),
	splitEvery(N,Y,T1).




% logBase2/2

logBase2(0,1).
logBase2(N,P):-
	logBase2(N,P,0).

logBase2(1,Acc,Acc). %Acc is used to make stuff easier
logBase2(N,P,Acc):-
	0 is N mod 2, %check if N is divisable by 2 always
	N1 is N//2,
	Acc1 is Acc +1,
	logBase2(N1,P,Acc1).

% getNumBits/4

getNumBits(_,fullyAssoc,_,0).
getNumBits(NumOfSets,setAssoc,Cache,BitsNum):-
	length(Cache,N),
	N1 is N//NumOfSets, % options in 1 set
	logBase2(N1,BitsNum).
getNumBits(_,directMap,Cache,BitsNum):-
	length(Cache,N),
	logBase2(N,BitsNum).



%fillZeros/4

fillZeros(S,N,R):-
	string_chars(S,L),
	fillZerosList(L,N,L2),
	atomics_to_string(L2,R).

fillZerosList(R,0,R).
fillZerosList([H|T],N,R):-
	N>0,
	N1 is N - 1,
	append(['0'],[H|T],R1),
	fillZerosList(R1,N1,R).

%replacement/8 Direct Mapping

replaceInCache(Tag,Idx,Mem,OldCache,NewCache,ItemData,
directMap,BitsNum):-
	number_string(Tag,S1), % String of Tag
	number_string(Idx,S2), % String of Idx
	string_to_list(S2,C1), % In case Idx had bin < BitsNum
	length(C1,N1),
	A is BitsNum - N1,
	fillZeros(S2,A,S22),
	string_concat(S1,S22,S3), % String of Address
	atom_number(S3,Address), % get Binary of address
	convertBinToDec(Address,AddInBin), % get decimal of address
	nth0(AddInBin,Mem,ItemData), % get ItemData at address
	string_to_list(S1,C2), % getting a list with length = N
	length(C2,N2), % N = number number of bits in current tag
	B is BitsNum + N2,
	Z is 6 - B, % getting number of zeros missing
	fillZeros(S1,Z,R), % Getting the full binary Tag
	convertBinToDec(Idx,ID), % Getting the decimal of index
	changeCache(R,ID,ItemData,OldCache,NewCache).

changeCache(Tag,0,ItemData,[_|T],[item(tag(Tag),data(ItemData),1,0)|T]).
changeCache(Tag,Idx,ItemData,[H1|T1],[H2|T2]):-
	Idx > 0,
	H1 = H2,
	Idx1 is Idx -1,
	changeCache(Tag,Idx1,ItemData,T1,T2).

%replacement/8 Fully-Associative

maximum(X,Y,X) :- X>=Y.
maximum(X,Y,Y) :- Y>X.

replaceInCache(Tag,0,Mem,OldCache,NewCache,ItemData,fullyAssoc,
_):-
	number_string(Tag,S1), % String of Tag
	atom_number(S1,Address), % get Binary of address
	convertBinToDec(Address,AddInBin), % get decimal of address
	nth0(AddInBin,Mem,ItemData), % get ItemData at address
	string_to_list(S1,C2), % getting a list with length = N
	length(C2,N), % N = number number of bits in current tag
	Z is 6 - N, % getting number of zeros missing
	fillZeros(S1,Z,R), % Getting the full binary Tag
	checkIfValid(OldCache,V),  % V acts as a checker as to which is used (Order or ValidBit)
	changeUsingValid(R,OldCache,NewCache,ItemData,V), % if V = 0 change using Valid
        getMaxOrder(OldCache,MaxOrder),
	changeUsingOrder(R,ItemData,MaxOrder,OldCache,NewCache,V). % if V = 1 change using Order


changeUsingValid(_,_,_,_,1).
changeUsingValid(Tag,[item(_,_,0,_)|T],[item(tag(Tag),data(ItemData),1,0)|T2],ItemData,0):-
	incrementOrder(T,T2).
changeUsingValid(Tag,[item(tag(Tag2),Data,ValidBit,Order)|T],[H2|T2],ItemData,0):-
	ValidBit == 1,
	Order2 is Order +1,
	H2 = item(tag(Tag2),Data,ValidBit,Order2),
	changeUsingValid(Tag,T,T2,ItemData,1).

incrementOrder([],[]).
incrementOrder([item(Tag,Data,0,Order)|T1],[item(Tag,Data,0,Order)|T2]):-
	incrementOrder(T1,T2).
incrementOrder([item(Tag,Data,1,Order)|T1],[item(Tag,Data,1,Order2)|T2]):-
	Order2 is Order + 1,
	incrementOrder(T1,T2).

getMaxOrder([],0).
getMaxOrder([item(_,_,_,Order)|T],M):-
	getMaxOrder(T,M1),
	maximum(Order,M1,M).

checkIfValid([],1).
checkIfValid([item(_,_,ValidBit,_)|T],V):-
	checkIfValid(T,V1),
	V is V1*ValidBit.

changeUsingOrder(_,_,_,_,_,0).
changeUsingOrder(Tag,ItemData,MaxOrder,[item(_,_,1,MaxOrder)|T],[item(tag(Tag),data(ItemData),1,0)|T2],1):-
	incrementOrder(T,T2).
changeUsingOrder(Tag, ItemData, MaxOrder, [item(tag(Tag2),Data,1,Order)|T], [H2|T2], 1):-
	Order \= MaxOrder,
	Order2 is Order + 1,
	H2 = item(tag(Tag2),Data,1,Order2),
	changeUsingOrder(Tag,ItemData,MaxOrder,T,T2,1).

replace(0, X, [_|T], [X|T]).
replace(N, X, [H|T],[H|T2]):-
	N > 0,
	N1 is N -1,
	replace(N1,X,T,T2).

%SET-ASSOCIATIVE

getDataFromCache(StringAddress,Cache,Data,HopsNum,setAssoc,SetsNum):-
	length(Cache,CacheLength),
	getNumBits(SetsNum,setAssoc,Cache,BitsNumNeeded),
	LengthPerSet is CacheLength / SetsNum,
	splitEvery(LengthPerSet,Cache,SplitCache),
	number_string(NumberAddress,StringAddress),
	convertAddress(NumberAddress,SetsNum,Tag,Idx,setAssoc),
	convertBinToDec(Idx,IdxDec),
	getHelperSet(Tag,SplitCache,Data,HopsNum,IdxDec,BitsNumNeeded).

getHelperSet(Tag,[H|T],Data,Hopsnum,0,BitsNumNeeded):-
	number_string(Tag,TagStringNoZeroes),
	string_length(TagStringNoZeroes,TagLength),
	N is 6-(TagLength+BitsNumNeeded),
	fillZeros(TagStringNoZeroes,N,StringTag),
	getHelperSet1(StringTag,H,Data,Hopsnum).

getHelperSet(Tag,[H|T],Data,Hopsnum,Index,BitsNumNeeded):-
	Index\=0,
	Index2 is Index-1,
	getHelperSet(Tag,T,Data,Hopsnum,Index2,BitsNumNeeded).

getHelperSet1(Tag,[H|T],Data,0):-
	H=item(tag(Tag),data(Data),1,_).

getHelperSet1(Tag,[H|T],Data,NumsBit):-
	H=item(tag(M),,,_),
	M\=Tag,
	getHelperSet1(Tag,T,Data,Nums2),
	NumsBit is (Nums2 + 1).

removeNumFromList(L,0,L,[]).
removeNumFromList([],N,L,M):-
	N\=0,
	N2 is N-1,
	removeNumFromList([],N2,L,M2),
	append(M2,['0'],M).
removeNumFromList([H|T],Num,L,M):-
	Num2 is (Num-1),
	removeNumFromList(T,Num2,L,M1),
	append(M1,[H],M).

convertAddress(Bin,SetsNum,Tag,Idx,setAssoc):-
	logBase2(SetsNum,I),
	number_string(Bin, BinString),
	string_chars(BinString, BinCharsL),
	reverse(BinCharsL,BinCharsL2),
	removeNumFromList(BinCharsL2,I,TagRevL,IdxL),
	reverse(TagRevL,TagL),
	string_chars(TagS,TagL),
	string_chars(IdxS,IdxL),
	checkIfZero(TagS,Tag),
	checkIfZero(IdxS,Idx).

checkIfZero(String,Number):-
	String = "",
	Number = 0 .

checkIfZero(String,Number):-
	String\="",
	number_string(Number,String).

reverse(Xs, Ys) :-
  reverse(Xs, [], Ys, Ys).
reverse([], Ys, Ys, []).
reverse([X|Xs], Rs, Ys, [_|Bound]) :-
  reverse(Xs, [X|Rs], Ys, Bound).

%replacement/8 Set-Associative


replaceInCache(Tag,Idx,Mem,OldCache,NewCache,ItemData,
setAssoc,SetsNum):-
	number_string(Tag,S1), % String of Tag
	number_string(Idx,S2), % String of Idx
	string_to_list(S2,C1), % In case Idx had bin < BitsNum
	length(C1,N1),
	logBase2(SetsNum,BitsNum),
	A is BitsNum - N1,
	fillZeros(S2,A,S22),
	string_concat(S1,S22,S3), % String of Address
	atom_number(S3,Address), % get Binary of address
	convertBinToDec(Address,AddInBin), % get decimal of address
	nth0(AddInBin,Mem,ItemData), % get ItemData at address
	string_to_list(S1,C2), % getting a list with length = N
	length(C2,N2), % N = number number of bits in current tag
	B is BitsNum + N2,
	Z is 6 - B, % getting number of zeros missing
	fillZeros(S1,Z,R), % Getting the full binary Tag String
	length(OldCache,N),
	Part is N // SetsNum,
	splitEvery(Part,OldCache,L),
	convertBinToDec(Idx,IdxDec),
	nth0(IdxDec,L,OldSet),
	checkIfValid(OldSet,V),
	changeUsingValid(R,OldSet,NewSet,ItemData,V),
	getMaxOrder(OldSet,MaxOrder),
	changeUsingOrder(R,ItemData,MaxOrder,OldSet,NewSet,V),
	replace(IdxDec,NewSet,L,L2),
	splitEvery(Part,NewCache,L2).

File: project.hs
data Item a = It Tag (Data a) Bool Int | NotPresent deriving (Show, Eq)
data Tag = T Int deriving (Show, Eq)
data Data a = D a deriving (Show, Eq)

data Output a = Out (a, Int) | NoOutput deriving (Show, Eq)
-------------------------------------------------------------------- GENERAL -----------------------------------------------------------

-- >> Converting a Binary number to a decimal number

convertBinToDec :: Integral a => a -> a
convertBinToDec b = convertBinToDecHelper b 0
convertBinToDecHelper 0 n = 0
convertBinToDecHelper b n = (b `mod` 10)*(2^n) + convertBinToDecHelper (div b 10) (n+1)

-- >> replaceIthItem

replaceIthItem :: (Eq a, Num a) => t -> [t] -> a -> [t]
replaceIthItem x (y:ys) 0 = [x] ++ ys
replaceIthItem x (y:ys) i = [y] ++ replaceIthItem x ys (i-1)

-- >> logBase2

logBase2 :: Floating a => a -> a
logBase2 n = logBase2Helper n 0
logBase2Helper 1 p = p
logBase2Helper n p = logBase2Helper (n/2) (p+1)

-- >> splitEvery

removeFirstN 0 l = l
removeFirstN n (x:xs) = (removeFirstN (n-1) xs)

splitEvery 0 l = [l]
splitEvery _ [] = []
splitEvery n l = (splitEveryHelper n l: splitEvery n (removeFirstN n l))

splitEveryHelper n [] = []
splitEveryHelper 0 l = []
splitEveryHelper n (x:xs) = (x:(splitEveryHelper (n-1) xs))																				

-- >> getNumBits

logBaseInt2 :: (Integral a, Num b) => a -> b
logBaseInt2 n = logBaseInt2Helper n 0
logBaseInt2Helper n i 
	| n<=1 = i
	| otherwise = logBaseInt2Helper (div n 2) (i+1)

getNumBits :: Num a => Int -> [Char] -> [b] -> a
getNumBits numOfSets cacheType cache 
	| cacheType == "fullyAssoc" = 0
	| cacheType == "setAssoc" = logBaseInt2(div (length cache) (numOfSets))
	| cacheType == "directMap" = (logBaseInt2 (length cache) )

-- >> fillZeros

fillZeros :: (Eq a, Num a) => [Char] -> a -> [Char]
fillZeros s 0 = s
fillZeros s n = "0" ++ fillZeros s (n-1)

------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------ COMPONENTS ------------------------------------------------------

-- >> getDataFromCache

getDataFromCache stringAddress cache "setAssoc" bitsNum = getDataFromCacheSetHelper (convertAddress (read stringAddress::Int) bitsNum "setAssoc") (splitEvery (2^bitsNum) cache) 

getDataFromCache address cache directMap bitsNum = getDataFromCachehelper (getDataFromCachegetindex (read address::Int) bitsNum) (getDataFromCachetag (read address::Int) bitsNum) cache

getDataFromCacheSetHelper (t,i) splitCache = getDataFromCacheSetHelper2 t (convertBinToDec i) splitCache

getDataFromCacheSetHelper2 tag 0 (x:xs) = getDataFromCacheSetHelper3 tag x 0
getDataFromCacheSetHelper2 tag index (x:xs) = getDataFromCacheSetHelper2 tag (index - 1) xs

getDataFromCacheSetHelper3 _ [] _ = NoOutput
getDataFromCacheSetHelper3 tag ((It (T t) (D a) (b) _ ):xs) hops 
	| (tag == t) && (b==True) = Out (a , hops)
	| otherwise = getDataFromCacheSetHelper3 tag xs (hops +1)

getFirstN 0 m = ""
getFirstN n (x:xs) = (x:(getFirstN (n-1) xs))

measureLength [] = 0
measureLength (x:xs) = 1 + measureLength xs 

checkIfZero "" = 0
checkIfZero n = read n :: Int

getDataFromCachehelper index tag cache = if b == True && t == tag then Out(d,tag) else NoOutput
										 where It (T t) (D d) b o = cache !! index
										
										

getDataFromCachegetindex address bitsNum= convertBinToDec(mod address (10 ^ bitsNum))
getDataFromCachetag address bitsNum= div address (10 ^ bitsNum)
-- >> convertAddress

convertAddress bin bitsNum "directMap"= ( div bin (10 ^ bitsNum), mod bin (10 ^ bitsNum))

convertAddress binAddress bitsNum "setAssoc" = ( (checkIfZero (getFirstN (measureLength (show binAddress) - bitsNum) (show binAddress) )) , (checkIfZero (reverse (getFirstN (bitsNum) (reverse (show binAddress))))))

convertAddress bin bitsNum "directMap"= ( div bin (10 ^ bitsNum), mod bin (10 ^ bitsNum))
----------------------------------------------------- REPLACEMENT ------------------------------------------------------

-- >> replaceInCache (directMap)

replaceInCache tag idx memory oldCache "directMap" bitsNum	= replaceHelperDirectMap tag (convertBinToDec idx) (getFromMemoryDM tag idx memory bitsNum) oldCache

-- >> replaceInCache (fullyAssoc)

replaceInCache tag idx memory oldCache "fullyAssoc" bitsNum	
						| isValidBit oldCache = ((getDataFromMemoryFA tag memory) ,replaceInCacheVB tag (getDataFromMemoryFA tag memory) oldCache)
						| otherwise = ((getDataFromMemoryFA tag memory) ,replaceInCacheMO tag (getDataFromMemoryFA tag memory) (getMaxOrder oldCache) oldCache)

-- >> replaceInCache (setAssoc)

replaceInCache tag idx memory oldCache "setAssoc" bitsNum
				| isValidBit (getSet idx oldCache bitsNum) = (getDataFromMemorySA tag idx memory bitsNum ,
				concat (replaceIthItem (replaceInCacheVB tag (getDataFromMemorySA tag idx memory bitsNum) (getSet idx oldCache bitsNum )) (splitEvery (2^bitsNum) oldCache)  (convertBinToDec idx)) )
				| otherwise = (getDataFromMemorySA tag idx memory bitsNum ,
				concat (replaceIthItem (replaceInCacheMO tag (getDataFromMemorySA tag idx memory bitsNum) (getMaxOrder (getSet idx oldCache bitsNum )) (getSet idx oldCache bitsNum )) (splitEvery (2^bitsNum) oldCache)  (convertBinToDec idx)) )



---------------------------------------------------------- Replacement Helpers ------------------------------------------

-- >> DirectMap Helpers
getFromMemoryDM tag idx memory bitsNum	| (length (show idx)) == bitsNum = memory!!(convertBinToDec (read (show tag ++ show idx)))
										| otherwise = memory!!(convertBinToDec (read (show tag ++ (fillZeros (show idx) (length (show idx))))))

replaceHelperDirectMap tag idx dat oldCache = (dat, replaceIthItem (It (T tag) (D dat) True 0) oldCache idx)

-- >> fullyAssoc Helpers

isValidBit [] = False
isValidBit ((It (T t) (D d) b o):xs) = if (b == False) then True else isValidBit xs

getDataFromMemoryFA tag memory = memory!!(convertBinToDec tag) -- returns the data specified by the address

replaceInCacheVB tag d ((It (T t) (D dat) b o):xs)	| b == False = ((It (T tag) (D d) True 0):incrementOrder xs) -- Replacing incase of valid bit
													| otherwise = ((It (T t) (D dat) b (o+1)): replaceInCacheVB tag d xs)

replaceInCacheMO tag d maxOrder ((It (T t) (D dat) b o):xs)	| o == maxOrder = ((It (T tag) (D d) True 0):incrementOrder xs) -- Replacing incase of no valid bit
															| otherwise = ((It (T t) (D dat) b (o+1)): replaceInCacheMO tag d maxOrder xs)

incrementOrder [] = [] -- incrementing the order of the rest of the valid items in cache
incrementOrder ((It (T t) (D dat) b o):xs)	| b == True = ((It (T t) (D dat) b (o+1)): incrementOrder xs)
											| otherwise = ((It (T t) (D dat) b (o)): incrementOrder xs)

getMaxOrder [] = 0 --  getting the max order (last item placed) in the cache
getMaxOrder ((It (T t) (D dat) b o):xs) = max o (getMaxOrder xs)

-- >> setAssoc Helpers

getSet idx oldCache bitsNum = (splitEvery (2^bitsNum) oldCache)!!(convertBinToDec idx)

getDataFromMemorySA tag idx memory bitsNum 	| (length (show idx)) == bitsNum = memory!!(convertBinToDec (read (show tag ++ show idx)))
											| otherwise = memory!!(convertBinToDec (read (show tag ++ (fillZeros (show idx) (length (show idx))))))


-- >> getData

getData stringAddress cache memory cacheType bitsNum
								| x == NoOutput = replaceInCache tag index memory cache cacheType bitsNum
								| otherwise = (getX x, cache)
				where
						x = getDataFromCache stringAddress cache cacheType bitsNum
						address = read stringAddress:: Int
						(tag, index) = convertAddress address bitsNum cacheType
						getX (Out (d, _)) = d

-- >> runProgram

runProgram [] cache _ _ _ = ([], cache)
runProgram (addr: xs) cache memory cacheType numOfSets = ((d:prevData), finalCache)
				where
							bitsNum = round(logBase2 numOfSets)
							(d, updatedCache) = getData addr cache memory cacheType bitsNum
							(prevData, finalCache) = runProgram xs updatedCache memory cacheType numOfSets